Katello/katello
# Fixes #33919 - support on_demand repo cleaning

#### What are the changes introduced in this pull request?
Adds support for:
1) Reclaiming space from a single repository
2) Reclaiming space from many repositories
3) Reclaiming space from an entire smart proxy

Reclaiming space means deleting any downloaded on-demand content in Pulp.  Only on-demand repositories are supported.

#### Considerations taken when implementing this change?
Most of it is straight-forward except for reclaiming space from a smart proxy. I integrated the space reclamation button into the existing sync dropdown.  I also used the same progress bar that is used during syncing.  I was considering just adding a separate button for reclaiming space, but I couldn't decide. Let me know what you think!

#### What are the testing steps for this pull request?
1) Create some on-demand and some immediate repos and sync at least the on-demand ones.
2) Download some on-demand RPMs from your repos' hosted contents
3) Clean contents via the UI (try bulk and single)
4) Check the Pulp task output to ensure the correct number of RPMs were cleaned up (look for "total: <num>" and "done: <num>")  
5) Sync a smart proxy and repeat steps 2-4 with the proxy.
  -> Clean the smart proxy by clicking the "Reclaim Space" button on the Synchronize drop-down.
6) Try cleaning only immediate repositories / smart proxies. It should fail with a warning.
7) Try cleaning a mix of immediate and on-demand repositories. It should succeed.

Questions:

~Would it be easy to know if the smart proxy's download policy in `capsule-content.controller.js`?  I suppose I could hide the clean option depending on how the smart proxy is configured.~
-> I fixed this by sneaking the download policy into the sync status view.

TODO:

- [x] Add warning about custom repositories
